demos/heat_eq_variable_source/setup.py

```diff
@@ -0,0 +1,67 @@
+from firedrake import *
+from firedrake_adjoint import *
+from pyroteus.metric import *
+from pyroteus.recovery import *
+from opt_adapt.opt import get_state
+
+
+def initial_mesh(n):
+    """
+    Construction of initial mesh
+    """
+    return RectangleMesh(5 * 2**n, 5 * 2**n, (-1, 1), (-1, 1))
+
+
+def initial_control(mesh):
+    """
+    The initial of control parameter
+    In this example, control parameter is not in Real space
+    """
+    V = FunctionSpace(mesh, "CG", 1)
+    return Function(V, name="Control")
+
+
+def forward_run(mesh, control, **kwargs):
+    """
+    Solve the PDEs in the given mesh
+    """
+    V = FunctionSpace(mesh, "CG", 1)
+    u = Function(V, name="State")
+    v = TestFunction(V)
+    m = Function(V, name="Control").assign(control)
+
+    # Run the forward model once to create the simulation record
+    F = (inner(grad(u), grad(v)) - m * v) * dx
+    bc = DirichletBC(V, 0.0, "on_boundary")
+    sp = {
+        "mat_type": "aij",  # Matrix type that we need to use a direct solver
+        "snes_monitor": None, # Print the nonlinear solver progress
+        "ksp_type": "preonly",  # Don't use a linear solver, just apply a preconditioner
+        "ksp_monitor": None,  # Print the linear solver progress
```

Comment:
I think once you are finished debugging, it would be good to drop the `"snes_monitor"` and `"ksp_monitor"` options.